File: sdks/kotlin/core/src/test/kotlin/io/getstream/visionagents/core/LiveTest.kt
package io.getstream.visionagents.core

import java.util.concurrent.CopyOnWriteArrayList
import kotlin.test.AfterTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNotEquals
import kotlin.test.assertNull
import kotlin.test.assertTrue
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.withTimeout
import kotlinx.serialization.json.jsonPrimitive
import org.junit.jupiter.api.condition.EnabledIfEnvironmentVariable
import kotlin.time.Duration.Companion.seconds
import kotlinx.coroutines.flow.first

/**
 * Tests that need a router running.
 *
 * The Kotlin answer to `@pytest.mark.integration`: skipped unless `VISION_AGENTS_URL` is set, so
 * the ordinary test run stays offline and fast.
 *
 *     VISION_AGENTS_URL=http://localhost:8080 VISION_AGENTS_CUSTOMER_ID=examples ./gradlew :core:test
 *
 * `VISION_AGENTS_AGENT` names an agent config to talk to; without one the router's defaults answer.
 */
@EnabledIfEnvironmentVariable(named = "VISION_AGENTS_URL", matches = ".+")
class LiveTest {
    private val agents = VisionAgents(
        url = System.getenv("VISION_AGENTS_URL"),
        customerId = System.getenv("VISION_AGENTS_CUSTOMER_ID") ?: "acme",
    )
    private val options = SessionOptions(agent = System.getenv("VISION_AGENTS_AGENT")?.ifEmpty { null })

    @AfterTest
    fun stop() {
        agents.close()
    }

    @Test
    fun `a text session joins no call and opens a socket`() = runBlocking {
        val chat = agents.chat(options)

        assertTrue(chat.session.isText)
        assertEquals("", chat.session.callId)
        assertEquals(Session.State.Live, chat.session.state)
        assertTrue(chat.isConnected.value)
        assertEquals(chat.id, agents.sessions.get(chat.id).id)
        chat.close()
    }

    @Test
    fun `asking something gets an answer, streamed into the transcript`() = runBlocking {
        val chat = agents.chat(options)

        chat.send("What is two plus two? Answer in one short sentence.")

        val done = withTimeout(30.seconds) {
            chat.conversation.first { it.state == Conversation.State.Idle && it.turns.size >= 2 }
        }
        assertTrue(done.turns.last().isAgent)
        assertTrue(done.turns.last().text.isNotEmpty())
        assertNull(chat.failure.value)
        chat.close()
    }

    @Test
    fun `a tool on this side is called and answered`() = runBlocking {
        val asked = CopyOnWriteArrayList<String>()
        val lookup = AgentTool(
            name = "lookup_order",
            description = "Look up one of the caller's orders by its order number.",
            parameters = AgentTool.strings(mapOf("order_id" to "the order number"), required = listOf("order_id")),
        ) { arguments ->
            asked += arguments["order_id"]?.jsonPrimitive?.content.orEmpty()
            "Order A-1042: 2 wool throws, 78.00, delivered 14 August, unopened."
        }
        val chat = agents.chat(options.copy(tools = listOf(lookup)))

        chat.send("Use the lookup_order tool to look up order A-1042 and tell me what is in it.")

        until(45) { asked.isNotEmpty() }
        assertEquals("A-1042", asked.first().uppercase())
        chat.close()
    }

    @Test
    fun `a rewound session carries on from the response kept, and a fork branches off it`() = runBlocking {
        val chat = agents.chat(options)

        chat.send("My name is Ada. Reply with one word.")
        withTimeout(30.seconds) { chat.conversation.first { it.state == Conversation.State.Idle && it.turns.size >= 2 } }
        chat.send("What is my name? Reply with one word.")
        withTimeout(30.seconds) { chat.conversation.first { it.state == Conversation.State.Idle && it.turns.size >= 4 } }
        until(10) { chat.responses.list().size == 2 }

        val kept = chat.responses.list().first()
        assertTrue(kept.said.contains("Ada"))
        assertTrue(chat.responses.items(kept.id).isNotEmpty())

        chat.responses.rewind(kept)
        assertEquals(listOf(kept.id), chat.responses.list().map { it.id })

        val fork = chat.fork(ForkOptions(responseId = kept.id))
        assertNotEquals(chat.id, fork.id)
        assertEquals(chat.id, fork.forkedFrom)
        agents.sessions.close(fork.id)
        chat.close()
    }

    @Test
    fun `a response created over http is a handle on a turn the socket then streams`() = runBlocking {
        val chat = agents.chat(options.copy(title = "kotlin live ${System.nanoTime()}"))

        val response = chat.responses.create("Say hello in one word.")

        assertEquals(chat.id, response.sessionId)
        withTimeout(30.seconds) { chat.conversation.first { it.turns.any { turn -> turn.isAgent && turn.text.isNotEmpty() } } }
        until(15) { chat.responses.list().any { it.id == response.id && it.status != AgentResponse.Status.Running } }
        chat.close()
    }

    @Test
    fun `a titled conversation is found by its title`() = runBlocking {
        val title = "kotlin search ${System.nanoTime()}"
        val session = agents.sessions.create(options.copy(title = title))

        val found = agents.sessions.search(title)

        assertTrue(found.any { it.id == session.id })
        assertTrue(agents.sessions.query(SessionQuery(limit = 50)).any { it.id == session.id })
        agents.sessions.close(session.id)
    }

    @Test
    fun `a guest can hold a conversation of their own`() = runBlocking {
        val guest = agents.guestUser(GuestOptions(name = "Kotlin guest"))
        assertTrue(guest.id.isNotEmpty() && guest.token.isNotEmpty())

        agents.setUser(guest)
        val session = agents.sessions.create(options)

        assertTrue(agents.sessions.query().any { it.id == session.id })
        agents.sessions.close(session.id)
    }
}

File: sdks/dart/core/test/live_test.dart
// Tests that need a router running.
//
// The Dart answer to @pytest.mark.integration: tagged `live`, and skipped unless
// VISION_AGENTS_URL is set, so the ordinary `dart test` stays offline and fast.
//
//     VISION_AGENTS_URL=http://localhost:8080 VISION_AGENTS_CUSTOMER_ID=examples \
//       VISION_AGENTS_AGENT=simple_voice_ai dart test -t live
@Tags(['live'])
library;

import 'dart:async';
import 'dart:io';

import 'package:test/test.dart';
import 'package:vision_agents_core/vision_agents_core.dart';

final _url = Platform.environment['VISION_AGENTS_URL'];
final _customerId = Platform.environment['VISION_AGENTS_CUSTOMER_ID'] ?? 'acme';

/// An agent config name. Empty opens a session with no config, which the router answers
/// with its own defaults.
final _agent = Platform.environment['VISION_AGENTS_AGENT'] ?? '';

/// A user of our own per run, so what one run lists is not another's.
final _user = 'dart-live-${DateTime.now().millisecondsSinceEpoch}';

void main() {
  late VisionAgents agents;
  final opened = <AgentSession>[];

  Future<AgentSession> chat([SessionOptions options = const SessionOptions()]) async {
    final session = await agents.agent(_agent).chat(options);
    opened.add(session);
    return session;
  }

  setUp(() {
    agents = VisionAgents(url: Uri.parse(_url!), customerId: _customerId).withUser(_user);
  });

  tearDown(() async {
    for (final session in opened) {
      await session.close();
    }
    opened.clear();
    agents.close();
  });

  group(
    'live',
    () {
      test('a text session joins no call and opens a socket', () async {
        final session = await chat();

        expect(session.session.isText, isTrue);
        expect(session.session.callId, isEmpty);
        expect(session.session.state, SessionState.live);
        expect(session.isConnected, isTrue);
      });

      test('asking something gets an answer, one delta at a time', () async {
        final session = await chat();
        final deltas = session
            .events()
            .where((event) => event.kind == AgentEventKind.responseDelta)
            .first;

        session.send('What are your opening hours? Answer in one sentence.');

        await deltas.timeout(const Duration(seconds: 30));
        await _until(() => session.state == ConversationState.idle && session.turns.length >= 2);
        expect(session.turns.last.speaker.isAgent, isTrue);
        expect(session.turns.last.text, isNotEmpty);
        expect(session.failure, isNull);
      });

      test('a tool on this side is called and answered', () async {
        final asked = <String>[];
        final session = await chat(
          SessionOptions(
            tools: [
              AgentTool(
                name: 'lookup_order',
                description: "Look up one of the caller's orders by its order number.",
                parameters: AgentTool.strings(
                  {'order_id': 'the order number'},
                  required: ['order_id'],
                ),
                run: (arguments) async {
                  asked.add('${arguments['order_id']}');
                  return 'Order A-1042: 2 wool throws, 78.00, delivered 14 August, unopened.';
                },
              ),
            ],
          ),
        );

        session.send('Look up order A-1042 and tell me what is in it.');

        await _until(() => asked.isNotEmpty, seconds: 45);
        expect(asked.first.toUpperCase(), 'A-1042');
        await _until(() => session.state == ConversationState.idle && session.turns.length >= 2);
        expect(session.turns.last.text.toLowerCase(), contains('throw'));
      });

      test('a rewound session carries on from the response kept, and forks from it', () async {
        final session = await chat();
        session.send('My name is Ada. Reply with one word.');
        await _until(() => session.state == ConversationState.idle && session.turns.length >= 2);
        session.send('What is my name? Reply with one word.');
        await _until(() => session.state == ConversationState.idle && session.turns.length >= 4);
        await _until(() async => (await session.responses.list()).length == 2);

        final kept = (await session.responses.list()).first;
        expect(kept.said, contains('Ada'));

        await session.rewind(kept.id);
        expect((await session.responses.list()).map((r) => r.id), [kept.id]);

        final fork = await session.fork(ForkOptions(responseId: kept.id));
        opened.add(fork);
        expect(fork.id, isNot(session.id));
        expect((await agents.sessions.get(fork.id)).forkedFrom, session.id);
      });

      test('a response created over HTTP is read back item by item', () async {
        final session = await chat();

        final response = await session.responses.create(
          'Say the word "pineapple" and nothing else.',
        );

        await _until(() async {
          final items = await session.responses.items(responseId: response.id);
          return items.any((item) => item.kind == ItemKind.answer);
        }, seconds: 45);
        final items = await session.responses.unwind(responseId: response.id).toList();
        expect(items.first.kind, ItemKind.said);
        expect(items.map((item) => item.text).join(' ').toLowerCase(), contains('pineapple'));
      });

      test('a conversation is found again by what it was called', () async {
        final title = 'dart live ${DateTime.now().microsecondsSinceEpoch}';
        final session = await chat(SessionOptions(title: title, project: 'dart-sdk'));

        await _until(() async => (await agents.sessions.search(title)).isNotEmpty);
        final found = await agents.sessions.search(title);
        final listed = await agents
            .agent(_agent)
            .sessions
            .query(const SessionQuery(project: 'dart-sdk'));

        expect(found.first.id, session.id);
        expect(listed.map((s) => s.id), contains(session.id));
        expect((await agents.sessions.get(session.id)).title, title);
      });

      test('closing a session ends it for the router too', () async {
        final session = await chat();

        await session.close();

        // Only a running session is found by id, so an ended one is looked for among the closed.
        await _until(() async {
          final closed = await agents.sessions.query(
            const SessionQuery(state: SessionFilter.closed, limit: 50),
          );
          return closed.any((s) => s.id == session.id);
        });
      });

      test('a guest holds a conversation of its own', () async {
        final store = MemoryGuestStore();
        final guest = await agents.guestUser(name: 'Dart Guest', store: store);
        expect(guest.id, startsWith('guest-'));
        expect(guest.token, isNotEmpty);
        expect(await agents.guestUser(store: store), guest);

        final asGuest = agents.withGuest(guest);
        final session = await asGuest.agent(_agent).chat();
        opened.add(session);

        expect((await asGuest.sessions.query()).map((s) => s.id), contains(session.id));
        asGuest.close();
      });

      test('looking something up answers with sources', () async {
        final found = await agents.router().search(
          'What is the capital of France?',
          const SearchOptions(results: 2),
        );

        expect(found.provider, isNotEmpty);
        expect(found.results, isNotEmpty);
      });
    },
    skip: _url == null ? 'set VISION_AGENTS_URL to run against a router' : false,
    timeout: const Timeout(Duration(seconds: 90)),
  );
}

/// Runs until the condition holds, so a test waits for what the model does rather than for a
/// fixed number of seconds.
Future<void> _until(FutureOr<bool> Function() done, {int seconds = 30}) async {
  final deadline = DateTime.now().add(Duration(seconds: seconds));
  while (!await done()) {
    if (DateTime.now().isAfter(deadline)) {
      fail('gave up after ${seconds}s');
    }
    await Future<void>.delayed(const Duration(milliseconds: 200));
  }
}
